pkg/jobmanager/job.go

```diff
@@ -74,45 +88,129 @@ func (j *Job) ExitCode() int {
+		slog.Error("Broadcaster closed with error.", "error", e1)
 	}
+	j.mu.Unlock()
+
+	// NOTE: cgroupPath is immutable and set at start before being shared, can
+	// safely be used without lock.
+	logger := slog.With("job_id", j.ID.String(), "cgroup_path", j.cgroupPath)
+
+	// Wait for ~1 second (arbitrary) for the cgroup to be empty.
+	const tickerInterval = 10 * time.Millisecond
+	const tickerAttempts = 100
+	ticker := time.NewTicker(tickerInterval)
+	defer ticker.Stop()
+	for range tickerAttempts {
+		empty, err := j.closeGroup()
+		if err != nil {
+			// Best effort.
+			logger.Error("Error closing group.", "error", err)
+			break
+		}
+		if empty {
+			if err := os.Remove(j.cgroupPath); err != nil {
+				// If cleanup failed, go over again freeze/kill/assert/cleanup.
+				// If it happens it likely means something violated the cgroup single writer principle.
+				logger.Error("Error removing cgroup on job close.", "error", err)
+			} else {
+				return
+			}
+		}
+		<-ticker.C
+	}
+	logger.Error("Timeout trying to cleanup cgroup.")
+}
+
+// wait for the underlying process. Broadcast the end via waitChan
+// and close the given resources.
+func (j *Job) wait() {
+	if err := j.cmd.Wait(); err != nil {
+		slog.Debug("Process Wait ended with error", "error", err)
+	}
+	j.close()
 }
 
 // NOTE: Expected to be called before being shared. Not locked.
 func (j *Job) start() error {
+	// Setup the cgroup limits.
+	cgroupDir, err := cgroups.New("job-" + j.ID.String())
+	if err != nil {
+		return fmt.Errorf("setup cgroups for job: %w", err)
+	}
+	defer func() {
+		// NOTE: This must be kept open until *after* the process started.
+		if err := cgroupDir.Close(); err != nil {
+			// Best effort. Either the process failed to start or started successfully and we don't need it anymore.
+			slog.Warn("Error closing cgroup from parent.", "error", err)
+		}
+	}()
+	// Make use of clone3 cgroup arg.
+	j.cmd.SysProcAttr.UseCgroupFD = true
+	j.cmd.SysProcAttr.CgroupFD = int(cgroupDir.Fd())
+	j.cgroupPath = cgroupDir.Name()
+
 	// Use the broadcaster as output for the process.
 	j.cmd.Stdout = j.broadcaster
 	j.cmd.Stderr = j.broadcaster // NOTE: Merge out/err for simplicity. Should split them for production.
 
+	// Control pipe.
+	r, w, err := os.Pipe()
+	if err != nil {
+		return fmt.Errorf("os.Pipe: %w", err)
+	}
+	j.cmd.ExtraFiles = []*os.File{w} // NOTE: We don't support setting extra files. Our pipe will always be '3'.
+
 	if err := j.cmd.Start(); err != nil {
 		// NOTE: We don't set a special status for 'failed to start' as this state
 		// will be discarded and garbage collected. Never surfaced to the user.
 		// When we implement listing, it may be interesting to add.
-		close(j.waitChan)
-		if e1 := j.broadcaster.Close(); e1 != nil {
-			// Best effort.
-			slog.Error("Broadcaster closed with error.", "error", e1)
-		}
-		return fmt.Errorf("start process: %w", err)
+		j.close()
+		return fmt.Errorf("start init process: %w", err)
```

Comment:
This early exit is not closing `w`, and nothing is closing `r`.

Reply:
Oops. Done.